### PR TITLE
Add changelog creator

### PR DESCRIPTION
**Pull request by MelanX**

### CODE REVIEW

File: src/main/java/org/moddingx/modlistcreator/Main.java [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
+import com.google.gson.GsonBuilder;
+import org.moddingx.modlistcreator.changelogger.Changelogger;
 import org.moddingx.modlistcreator.modlist.ModListCreator;
 
 import java.io.IOException;
 import java.util.Locale;
 
 public class Main {
+
+    public static Gson GSON;
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/Main.java]: Should be `final`

File: src/main/java/org/moddingx/modlistcreator/changelogger/ChangelogFormatter.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ChangelogFormatter {
+
+    public static String format(Modpack from, Modpack to, OutputTarget.Type outputType) {
+        Map<String, Modpack.File> oldInfoBySlug = from.files().stream().collect(Collectors.toMap(Modpack.File::projectSlug, info -> info));
+        Map<String, Modpack.File> newInfoBySlug = to.files().stream().collect(Collectors.toMap(Modpack.File::projectSlug, info -> info));
+
+        if (from.getClass() != to.getClass()) {
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/changelogger/ChangelogFormatter.java]: Is this really needed?

File: src/main/java/org/moddingx/modlistcreator/changelogger/Changelogger.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import java.nio.file.StandardOpenOption;
+import java.util.List;
+
+public class Changelogger {
+
+    public static void run(String[] args) throws IOException {
+        OptionParser options = new OptionParser();
+        OptionSpec<Path> specOld = options.acceptsAll(List.of("o", "old"), "Defines the old modpack zip or json file").withOptionalArg().withValuesConvertedBy(new PathConverter(PathProperties.FILE_EXISTING)).defaultsTo(Paths.get("old.json"));
+        OptionSpec<Path> specNew = options.acceptsAll(List.of("n", "new"), "Defines the new modpack zip or json file").withOptionalArg().withValuesConvertedBy(new PathConverter(PathProperties.FILE_EXISTING)).defaultsTo(Paths.get("new.json"));
+        OptionSpec<String> specOutput = options.accepts("output", "Defines the output file name without extension").withOptionalArg().ofType(String.class).defaultsTo("changelog");
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/changelogger/Changelogger.java]: For full file names, extension should be part of the file name as well. (Is a default value even needed here?)

File: src/main/java/org/moddingx/modlistcreator/changelogger/Changelogger.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            Path from = set.valueOf(specOld);
+            Path to = set.valueOf(specNew);
+
+            if (!Files.exists(from)) {
+                throw new IllegalStateException("File does not exist: " + from);
+            }
+
+            if (!Files.exists(to)) {
+                throw new IllegalStateException("File does not exist: " + to);
+            }
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/changelogger/Changelogger.java]: Already checked by the option parser

File: src/main/java/org/moddingx/modlistcreator/changelogger/Changelogger.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+            OutputTarget.Type outputType = set.valueOf(specFormat);
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/changelogger/Changelogger.java]: The option spec allows multiple values separated with `,` but you only query one of them.

File: src/main/java/org/moddingx/modlistcreator/changelogger/Changelogger.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            Path output = Paths.get(set.valueOf(specOutput) + "." + outputType.extension);
+            Files.writeString(output, ChangelogFormatter.format(CommonUtils.fromPath(from), CommonUtils.fromPath(to), outputType), StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
+            System.exit(0);
+        } catch (OptionException e) {
+            System.err.println(e.getMessage() + "\n");
+            options.printHelpOn(System.err);
+            System.exit(1);
+            throw new Error();
+        }
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/changelogger/Changelogger.java:41]: Unnecessary large `try`-`catch` block, see how `ModListCreator.java` does this.

File: src/main/java/org/moddingx/modlistcreator/changelogger/README.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# Changelog
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/changelogger/README.md]: Please move to a `docs` directory in repository root, just as [ModGradle](https://github.com/ModdingX/ModGradle/tree/master/docs) does.

File: src/main/java/org/moddingx/modlistcreator/modlist/README.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
+# Modlist
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/modlist/README.md]: Please move to a `docs` directory in repository root, just as [ModGradle](https://github.com/ModdingX/ModGradle/tree/master/docs) does.

File: src/main/java/org/moddingx/modlistcreator/output/HtmlTarget.java [hunk 9]
```diff
@@ -15,6 +15,11 @@ public void addHeader(String content) {
         this.tag.appendChild(new Element("h2").append(content));
     }
 
+    @Override
+    public void addSubHeader(String content) {
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/output/HtmlTarget.java:19]: `h3` to match markdown

File: src/main/java/org/moddingx/modlistcreator/util/CommonUtils.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+import org.moddingx.modlistcreator.output.OutputTarget;
+import org.moddingx.modlistcreator.platform.Modpack;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.ProviderNotFoundException;
+import java.util.function.BiFunction;
+
+public class CommonUtils {
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/util/CommonUtils.java]: Maybe name it `OptionUtil` ?

File: src/main/java/org/moddingx/modlistcreator/util/CommonUtils.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    public static Modpack fromPath(Path path) throws IOException {
+        try {
+            return Modpack.loadZip(path);
+        } catch (ProviderNotFoundException e) {
+            // Not a zip file
+            return Modpack.load(path);
+        }
+    }
```
Comment (noeppi-noeppi) [inline src/main/java/org/moddingx/modlistcreator/util/CommonUtils.java]: Could be moved to `Modpack` class

**Review by noeppi-noeppi — CHANGES_REQUESTED** [review]

File: src/main/java/org/moddingx/modlistcreator/changelogger/Changelogger.java (hunk 3 above)
Comment (MelanX) [reply src/main/java/org/moddingx/modlistcreator/changelogger/Changelogger.java]: You can't define the correct file extension if you select multiple output types. And I'd say yes, a default name is quiet useful.

**Pull request merged**